### PR TITLE
Fix course identifiers

### PR DESCRIPTION
**Pull request by marianysilva**
# Description

- [x] Add Course ID `GENERATED ALWAYS AS IDENTITY`
- [x] Rename `id` params to `courseUUID``
- [x] Improve Course SQL queries
- [x] Remove Update By Code

## Type of change

- [ ] 🐞 Bugfix _(fixes an issue)_
- [ ] 🆕 New feature _(adds functionality)_
- [x] 🚀 Code Enhancement _(improve code quality)_
- [ ] 🖌️ UI Improvement _(improves UI/UX experience)_
- [ ] ⏺️ Migration _(schema update and/or data migration)_
- [ ] ⚙️ Dependency Update _(dependencies have been updated)_
- [ ] 🗃 Documentation Update _(adds/updates the documentation)_

# How Has This Been Tested?

- [x] `go mod vendor`
- [x] `go get ./...`
- [x] `make test`
- [x] `make migration-up`
- [x] `make migration-down`

## Summary by CodeRabbit

- **New Features**
	- Enhanced the courses database schema for better integrity and efficiency.
- **Refactor**
	- Improved SQL query structure and naming conventions for course operations.
	- Shifted from using integer IDs to UUIDs for course identification across the application.
- **Bug Fixes**
	- Fixed unique constraint issues by adjusting the `courses` table structure.
- **Documentation**
	- Updated internal documentation to reflect changes in parameter naming and method signatures.

### CODE REVIEW

File: internal/course/database/course_queries.go
Comment (rluders) [inline internal/course/database/course_queries.go:8]: Wrong code style, rename it to `returningColumns` and add a comment to explain where and for what it is being used.

File: internal/course/database/course_queries.go
Comment (rluders) [inline internal/course/database/course_queries.go:15]: Well, if `getCourse` also gets it by uuid and doesn't have it explicitly on its name, I would probably rename it to `updateCourse`.

File: internal/course/database/course_repository.go
Comment (rluders) [inline internal/course/database/course_repository.go:90]: `UpdateCourse`

File: internal/course/database/course_repository_test.go
Comment (rluders) [inline internal/course/database/course_repository_test.go:208]: `TestRepository_UpdateCourse`

File: internal/course/domain/course.go
Comment (rluders) [inline internal/course/domain/course.go:14]: The fields: `Image`, `ImageCover`, and `Description` are nullable in the database, in this case, we need to add the `omitempty` to the JSON annotation.

File: internal/course/domain/course_repository.go
Comment (rluders) [inline internal/course/domain/course_repository.go:9]: `UpdateCourse`

File: internal/course/domain/course_service.go
Comment (rluders) [inline internal/course/domain/course_service.go:37]: `UpdateCourse`

**Review by rluders — CHANGES_REQUESTED** [review]

File: internal/course/domain/course_repository.go
Comment (marianysilva) [reply internal/course/domain/course_repository.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/domain/course.go
Comment (marianysilva) [reply internal/course/domain/course.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/database/course_repository_test.go
Comment (marianysilva) [reply internal/course/database/course_repository_test.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/database/course_repository.go
Comment (marianysilva) [reply internal/course/database/course_repository.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go]: https://github.com/sumelms/microservice-course/pull/19/commits/6244cbf2156cfdfa417826cf5017f6c73799dd89

File: internal/course/database/course_queries.go
Comment (marianysilva) [inline internal/course/database/course_queries.go:9]: Shoud we return delet_at column? @rluders

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [inline db/migrations/1657379568_create_courses_table.up.sql:5]: Keep bigserial here

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [inline db/migrations/1657379568_create_courses_table.up.sql:7]: shouldn't it be unique?, there shouldn't be 2 courses with the same codes

**Review by arianerocha — CHANGES_REQUESTED** [review]

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [inline db/migrations/1657379568_create_courses_table.up.sql:19]: Course not subscription

**Review by arianerocha — CHANGES_REQUESTED** [review]

File: internal/course/database/course_repository.go
Comment (arianerocha) [inline internal/course/database/course_repository.go:114]: I still think it makes sense to have update by code. But also depends how we planned to expose the list of courses.

File: internal/course/database/course_queries.go
Comment (arianerocha) [inline internal/course/database/course_queries.go:29]: If we plan on limiting this by active/archived, I suggest creating a `get_active` and `get_archived`, or you can remove the delete_at null and let this be populated via parameter

File: internal/course/database/course_queries.go
Comment (arianerocha) [reply internal/course/database/course_queries.go:9]: Yep

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [inline db/migrations/1657379568_create_courses_table.up.sql:24]: With this we will never be able to unarchive a course, do we really want to do that?

**Review by arianerocha — CHANGES_REQUESTED** [review]

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (rluders) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: Yeah, it should be unique for sure.

File: internal/course/database/course_repository.go
Comment (rluders) [reply internal/course/database/course_repository.go:114]: I think that for now, we can help it simple. We can include these extra scenarios later. There would be also the need to get by code, but... let's first achieve the basics.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:19]: I added a unique index based on the code and a "deleted_at" field because I anticipate that we might have duplicate codes as long as one of them has been marked as deleted:

```
CREATE UNIQUE INDEX courses_code_uindex
    ON courses (code, deleted_at) NULLS NOT DISTINCT;
```

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go:9]: As we are using the "deleted_at" column for soft deletion, I believe that our CRUD queries should specifically filter for records where "deleted_at" is null. This ensures that we only interact with active records during regular operations, maintaining the integrity and relevance of our data. Additionally, I propose that we create a dedicated endpoint to retrieve all records, including those marked as deleted. This endpoint would include the "deleted_at" information in its response, allowing for a comprehensive view of the data, both active and deleted. This approach enables us to keep the regular data interactions clean and focused while still providing the capability to audit or review the entire dataset when needed. What do you think?

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:5]: `id bigserial PRIMARY KEY GENERATED ALWAYS AS IDENTITY` seems to combine elements from two different approaches to generating unique identifiers in PostgreSQL.

We can use the data type `bigint`, which is designed to store very large integers. It can hold values from -9223372036854775808 to 9223372036854775807.

This choice is suitable for tables expected to grow beyond the capacity of the standard integer type, which is limited to a range of about -2 billion to +2 billion.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: https://github.com/sumelms/microservice-course/pull/19#discussion_r1520508615

Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: I added a unique index based on the code and a "deleted_at" field because I anticipate that we might have duplicate codes as long as one of them has been marked as deleted:

```
CREATE UNIQUE INDEX courses_code_uindex
    ON courses (code, deleted_at) NULLS NOT DISTINCT;
```

Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: https://github.com/sumelms/microservice-course/pull/19#discussion_r1520508615

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:19]: Using soft delete means we can only update records that have not been deleted. If someone wants to modify a deleted record, they must first reactivate the record, then make the desired changes, and finally, if necessary, delete it again. This method preserves the integrity of historical data and allows for the recovery or analysis of deleted entries. However, it also requires additional logic in the application to distinguish between active and deleted records, especially during update operations. The process of reactivating, modifying, and optionally re-deleting a record might seem cumbersome but it's essential for maintaining data consistency and adhering to the soft delete pattern.

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go:29]: https://github.com/sumelms/microservice-course/pull/19#discussion_r1520508615

https://github.com/sumelms/microservice-course/pull/19#discussion_r1520517473

File: internal/course/database/course_queries.go
Comment (arianerocha) [reply internal/course/database/course_queries.go:9]: I don't see the need for 2 endpoints, I believe it could be filtered data.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql:5]: go for bigint then

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: This can prevent for us to unarchive the information, I prefer to keep the possibilty, specially if someone archive by mistake.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql:19]: I know what soft delete is, but we need to decide whether deleted_at will have the archived behavior or not. We can change archived data, by unarchiving for example.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: @rluders Can you help us here?

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go:29]: @rluders @arianerocha  additional comments?

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:19]: If we are aligned, I will just update our comment on this column.

File: internal/course/database/course_queries.go
Comment (marianysilva) [reply internal/course/database/course_queries.go:9]: I'm going to create an issue to discuss and implement solutions related to recovering and listing deleted data, ok? @arianerocha @rluders

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql]: https://github.com/sumelms/microservice-course/pull/19/commits/c04e66a4cce0b4bf28f7e13db39c5d4d1c41e173

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql]: https://github.com/sumelms/microservice-course/pull/19/commits/c04e66a4cce0b4bf28f7e13db39c5d4d1c41e173

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: yep, It may seem logical, but it is always good to consider what we are determining as a business rule, and not just the user experience. In this format, if someone archives a course, even by mistake, it will not be possible to unarchive it if someone enters the same code, inactivating all relationships and data related to that course. If the user wants to consult the course history, which means consulting archived courses, this could mean a list of courses without code distinctions and without knowing why that course was archived. If we are considering these factors, it is okay to continue like this.

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql]: Just remove the historical part in the description the rest is great

File: db/migrations/1657379568_create_courses_table.up.sql
Comment (arianerocha) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: we may add this as part of the archive/unarchive behivior discussuion to no continues to block this PR, WDYT?

Comment (marianysilva) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: @rluders

Comment (rluders) [reply db/migrations/1657379568_create_courses_table.up.sql:7]: Agreed.

**Review by rluders — APPROVED** [review]
lgtm

**Comment by rluders** [comment]
@arianerocha since you where involved in the discussions, I'll wait for your approval before squash and merge. :)

**Review by arianerocha — APPROVED** [review]

**Pull request merged**
